tinygrad/spec.py

```diff
@@ -1,7 +1,18 @@
-from typing import cast
-from tinygrad.ops import PatternMatcher, UPat, GroupOp, Ops, UOp, print_uops
+from typing import cast, Callable
+from tinygrad.ops import PatternMatcher, UPat, GroupOp, Ops, UOp, print_uops, python_alu
 from tinygrad.dtype import DType, ImageDType, dtypes, PtrDType
-from tinygrad.helpers import all_same, dedup, prod, getenv, DEBUG
+from tinygrad.helpers import all_same, dedup, flatten, prod, getenv, DEBUG, CHECK_OOB
+if CHECK_OOB:
+  import z3
+
+  # IDIV is truncated division but z3 does floored division and mod by power of two sometimes uses Ops.AND
+  def z3_cdiv(a,b): return z3.If(a<0, (a+(b-1))/b, a/b)
+  z3_alu: dict[Ops, Callable] = python_alu | {Ops.MOD: lambda a,b: a-z3_cdiv(a,b)*b, Ops.IDIV: z3_cdiv, Ops.CAST: lambda x:x,
+    Ops.SHR: lambda a,b: a/(2**b), Ops.SHL: lambda a,b: a*(2**b), Ops.AND: lambda a,b: a%(b+1) if isinstance(b, (int, z3.ArithRef)) else a&b}
+  def z3_eval(u: UOp|int, ctx: dict[UOp, 'z3.ArithRef']) -> 'int|z3.ArithRef':
```

Comment:
Can this be done with a graph_rewrite? Like look at how we do rendering.

The problem with this is that it isn't cached and you can get 2^n blowup.